GDScript: Change default value for untyped exported variables

dalexeev opened the pull request:
Closes #74188.

See also comments in #74231 and #75555.

Formally, this breaks compatibility, but only in a small number of cases, and we assume that users expect this behavior rather than the current one. However, it needs to be documented anyway.

dalexeev committed:
GDScript: Change default value for untyped exported variables

vnen commented:
Reviving this, do we want to move forward? I wonder what's the solution for invalid default values (e.g. when `0` is outside the range or does not belong to the define enum). AFAIK the inspector still hasn´t been fixed for those cases.

Technically, since the annotation arguments are constants we could pick a valid default, but I'm not sure if it's worth the trouble.

dalexeev commented:
> Reviving this, do we want to move forward? I wonder what's the solution for invalid default values (e.g. when `0` is outside the range or does not belong to the define enum). AFAIK the inspector still hasn´t been fixed for those cases.

I think this should be resolved only on the inspector's side.

![](https://github.com/godotengine/godot/assets/47700418/a9018ec0-fa02-4b7b-98cc-046904e8a22c)
![](https://github.com/godotengine/godot/assets/47700418/a91a3e3a-a0e6-4e34-93c7-7ff46ff122ca)

* Also, I think we're a little late with this PR. I made it because you suggested this option instead of #74231. But as an alternative, we could make a warning instead of an error. This would be consistent with the recently merged #90756.

KoBeWi commented:
I closed the issue, as it was resolved differently. I think this PR no longer needed?

dalexeev commented:
* This PR is outdated, and I wasn't initially a supporter of this approach. I implemented it after vnen's comment regarding #74231. But it's also no longer applicable, since we've added the `Variant` export. So, at present, I think the right approach is similar to #112772, as suggested by akien-mga.
